public/app/features/live/live.ts

```diff
@@ -95,6 +93,77 @@ class CentrifugeSrv implements GrafanaLiveSrv {
     console.log('onError', context);
   };
 
+  /**
+   * Returns a channel that has been initalized and subscribed to the given path
+   */
+  private async getChannel<T>(pluginId: string, path: string): Promise<Channel<T>> {
+    const key = `${pluginId}/${path}`;
+    const c = this.channels.get(key);
+    if (c) {
+      return Promise.resolve(c);
+    }
+
+    const liveSupport = await this.getChannelSupport(pluginId);
+    const handler = liveSupport.getChannelHandler(path);
+    if (!handler) {
+      return Promise.reject(`Invalid path: ${key}`);
+    }
+
+    // Make sure we are connected before trying to subscribe
+    if (!this.isConnected()) {
+      await new Promise<void>(resolve => {
+        if (this.centrifuge.isConnected()) {
+          return resolve();
+        }
+        const connectListener = () => {
+          resolve();
+          this.centrifuge.removeListener('connect', connectListener);
+        };
+        this.centrifuge.addListener('connect', connectListener);
+      });
+    }
+
+    const subject = new Subject<T>();
+    const callbacks: SubscriptionEvents = {
+      ...this.standardCallbacks,
+      publish: (ctx: PublicationContext) => {
+        // console.log('GOT', JSON.stringify(ctx.data), ctx);
+        const v = handler!.onMessageReceived(ctx.data);
+        subject.next(v);
+      },
+    };
+    console.log('initChannel', this.centrifuge.isConnected(), key, handler);
+    const subscription = this.centrifuge.subscribe(key, callbacks);
+    const channel = {
+      plugin: pluginId,
+      path,
+      subject,
+      stream: subject.pipe(
+        finalize(() => {
+          console.log('Final listener for', key);
+          if (subscription) {
+            subscription.unsubscribe();
+          }
+          this.channels.delete(key); // remove the listener when done
+        }),
+        share()
+      ),
+      handler,
+      subscription,
+    };
+    this.channels.set(key, channel);
+    return channel;
+  }
+
+  private async getChannelSupport(pluginId: string): Promise<LiveChannelSupport> {
+    return loadPlugin(pluginId).then(plugin => {
+      if (!plugin.liveSupport) {
+        return Promise.reject('Plugin does not have live support configured');
+      }
+      return plugin.liveSupport;
+    });
```

Comment:
Personal preference, but here's the function with async/await syntax

```suggestion
    const plugin = await loadPlugin(pluginId);
    if (!plugin.liveSupport) {
      throw new Error('Plugin does not have live support configured');
    }

    return plugin.liveSupport;
    });
```